cpp/src/arrow/array/concatenate_test.cc

```diff
@@ -539,4 +539,13 @@ TEST_F(ConcatenateTest, OffsetOverflow) {
   ASSERT_RAISES(Invalid, Concatenate({fake_long, fake_long}).status());
 }
 
+TEST_F(ConcatenateTest, DictionaryConcatenateWithEmptyUint16) {
+  auto dict_type = dictionary(uint16(), utf8());
+  auto dict_one = DictArrayFromJSON(dict_type, "[]", "[]");
+  auto dict_two = DictArrayFromJSON(dict_type, "[0, 1, null]", "[\"A0\", \"A1\"]");
+  ASSERT_OK_AND_ASSIGN(auto concat_actual, Concatenate({dict_one, dict_two}));
```

Comment:
nit: it might be worth concatenating with a run of a few nulls?

Reply:
good suggestion, makes sense. Done in commit: 204981dc56a71b628286a9e9264ad634a3c90d8d